Sources/BSON/BSONBinary.swift

```diff
@@ -117,6 +117,72 @@ public struct BSONBinary: Equatable, Hashable {
         try self.init(data: dataObj, subtype: subtype)
     }
 
+    /*
+     * Initializes a `Binary` from ExtendedJSON.
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for a `Binary`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value does not conform to the `Binary` syntax.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        switch json {
+        case let .object(obj):
+            // canonical and relaxed extended JSON
+            guard let binary = obj["$binary"] else {
+                return nil
+            }
+            guard obj.count == 1 else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected only \"$binary\", found too many keys: \(obj.keys)"
+                )
+            }
+            guard let binaryObj = binary.objectValue,
+                binaryObj.count == 2,
+                let base64 = binaryObj["base64"],
+                let subTypeInput = binaryObj["subType"] else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected \"base64\" and \"subType\" keys in the object at \"$binary\", " +
+                        "found keys: \(obj.keys)"
+                )
+            }
+            guard let base64Str = base64.stringValue else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Could not parse `base64` from \"\(base64)\", " +
+                        "input must be a base64-encoded (with padding as =) payload as a string"
+                )
+            }
+            guard let subTypeStr = subTypeInput.stringValue,
+                let subTypeInt = UInt8(subTypeStr),
```

Comment:
since the subtype is stored as a hex value, I think we need `UInt8(subTypeStr, radix: 16)`. Otherwise, it may assume it's initializing a decimal value.